### PR TITLE
Fix bugs with concurrent use of MEMORY ccaches

### PR DESCRIPTION
**Pull request by greghudson**
[This bug is serious enough to deserve a backport, but the fix is more complicated than I'm normally comfortable backporting.  I will probably decide to backport it but I'm still considering.]

A memory ccache iterator stores an alias into the cache object's
linked list of credentials.  If the cache is reinitialized while the
iterator is active, the alias becomes invalid.  Also, multiple handles
referencing the same memory ccache all use aliases to the same data
object; if one of the handles is destroyed, the other contains a
dangling pointer.

Fix the first issue by adding a generation counter to the cache and to
cursors, incremented each time the cache is initialized or destroyed.
Check the generation on each cursor step and end the iteration if the
list was invalidated.  Fix the second issue by adding a reference
count to the cache object, counting one reference for the table slot
and one for each open handle.  Empty the cache object on each destroy
operation, but only release the object when the last handle to it is
destroyed or closed.

Add regression tests for the two issues to t_cc.c.

The first issue was reported by Sorin Manolache.

ticket: 8707

### CODE REVIEW

**Comment by greghudson** [comment]
Updated to keep the cache locked until krb5_mcc_next_cred() is done; otherwise the pointer in the cursor could be freed by another thread (via krb5_cc_initialize() or krb5_cc_destroy() on another handle referencing the same cache) after the generation is checked.  This race wasn't initially obvious to me because the code after the generation check doesn't appear the access the cache, only the cursor, so I opted for an explicit comment.

**Comment by frozencemetery** [comment]
I think the ticket number is wrong - [8707 is the ulog resource leak](http://krbdev.mit.edu/rt/Ticket/Display.html?id=8707).  I think it should be [8202](http://krbdev.mit.edu/rt/Ticket/Display.html?id=8202).

File: src/lib/krb5/ccache/cc_memory.c [hunk 1]
```diff
@@ -185,39 +205,34 @@ krb5_mcc_initialize(krb5_context context, krb5_ccache id, krb5_principal princ)
...
- *
- * Errors:
- * system errors (locks related)
  */
 krb5_error_code KRB5_CALLCONV
 krb5_mcc_destroy(krb5_context context, krb5_ccache id)
 {
     krb5_mcc_list_node **curr, *node;
-    krb5_mcc_data *d;
+    krb5_mcc_data *d = id->data;
```
Comment (frozencemetery) [inline src/lib/krb5/ccache/cc_memory.c]: This is duplicated five lines below (right before the `for` loop).

**Comment by greghudson** [comment]
Fixed both issues.

**Pull request merged**